File: testdata/fixtures/go/CWE-252.go
package fixture

// CWE-252: the error from a random read is thrown away.
//
// crypto/rand.Read fills a buffer and returns how much it filled. If it
// fails - and it can, on a machine whose entropy source is not ready or whose
// file descriptors have run out - the buffer keeps whatever it had, which for
// a freshly made slice is zeros. A key, a token or an IV made this way is a
// constant, and nothing about the program says so afterwards: it runs, it
// encrypts, and every value it produces is the same one.
//
// Go will not let the return be ignored silently, so the two spellings that
// throw it away are the two this reports: a bare call, and an assignment to
// the blank identifier.

import (
	"crypto/rand"
	"io"
)

func tokenFromIgnoredRead() []byte {
	b := make([]byte, 32)
	// ruleid: go-unchecked-random-read
	rand.Read(b)
	return b
}

func keyFromDiscardedError() []byte {
	key := make([]byte, 32)
	// ruleid: go-unchecked-random-read
	_, _ = rand.Read(key)
	return key
}

func ivFromDiscardedFullRead() []byte {
	iv := make([]byte, 16)
	// ruleid: go-unchecked-random-read
	_, _ = io.ReadFull(rand.Reader, iv)
	return iv
}

// The error is checked, which is the whole difference.
func checkedToken() ([]byte, error) {
	b := make([]byte, 32)
	// ok: go-unchecked-random-read
	if _, err := rand.Read(b); err != nil {
		return nil, err
	}
	return b, nil
}

// Assigned and returned, so the caller decides.
func returnedError() ([]byte, error) {
	b := make([]byte, 32)
	// ok: go-unchecked-random-read
	n, err := rand.Read(b)
	_ = n
	return b, err
}

// Handed straight back to the caller.
func passesItOn(b []byte) (int, error) {
	// ok: go-unchecked-random-read
	return rand.Read(b)
}

File: testdata/fixtures/c/CWE-252.c
/* CWE-252: Unchecked Return Value.
 *
 * read() returns how much it read, which is not how much was asked for. A
 * program that ignores it uses a buffer that was filled partly, or not at
 * all, with whatever was in it before - and on a socket "not at all" is the
 * ordinary case rather than the error case.
 */
#include <stdio.h>
#include <string.h>
#include <sys/socket.h>
#include <unistd.h>

int load(int fd, int sock, FILE *f, char *out, size_t n)
{
    char header[16];
    char line[128];

    /* ruleid: c-unchecked-read */
    read(fd, header, sizeof header);
    /* ruleid: c-unchecked-read */
    recv(sock, out, n, 0);
    /* ruleid: c-unchecked-read */
    fread(out, 1, n, f);
    /* ruleid: c-unchecked-read */
    fgets(line, sizeof line, f);
    /* ruleid: c-unchecked-read */
    (void)pread(fd, out, n, 0);

    /* ok: c-unchecked-read */
    if (read(fd, header, sizeof header) != (ssize_t)sizeof header)
        return -1;
    /* ok: c-unchecked-read */
    ssize_t got = recv(sock, out, n, 0);
    if (got <= 0)
        return -1;
    /* ok: c-unchecked-read */
    while (fread(out, 1, n, f) == n)
        continue;
    /* ok: c-unchecked-read */
    if (fgets(line, sizeof line, f) == NULL)
        return -1;
    /* ok: c-unchecked-read */
    return (int)read(fd, out, n);
}
